pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,459 @@
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON headers: %w", err)
+	}
+
+	var payload []byte
+	if len(detachedPayload) == 0 {
+		payload, err = base64.RawURLEncoding.DecodeString(parts[jwtPayloadPart])
+		if err != nil {
+			return nil, fmt.Errorf("decode base64 payload: %w", err)
+		}
+	} else {
+		payload = detachedPayload
+	}
+
+	var jwtPayload map[string]interface{}
+
+	err = json.Unmarshal(payload, &jwtPayload)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON payload: %w", err)
+	}
+
+	signature, err := base64.RawURLEncoding.DecodeString(parts[jwtSignaturePart])
+	if err != nil {
+		return nil, fmt.Errorf("decode base64 signature: %w", err)
+	}
+
+	jwt := &jsonWebToken{
+		headers:   jwtHeaders,
+		payload:   jwtPayload,
+		signature: signature,
+	}
+
+	if verifier != nil {
+		err = verifier.Verify(jwt)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return jwt, nil
+}
+
+// FromUnsecured parses unsecured JWT in compact serialization format.
+func FromUnsecured(jwtUnsecured string, detachedPayload []byte) (JSONWebToken, error) {
+	parts := strings.Split(jwtUnsecured, ".")
+	if len(parts) != jwtPartsCount {
+		return nil, errors.New("invalid JWT compact format")
+	}
+
+	headers, err := base64.RawURLEncoding.DecodeString(parts[jwtHeaderPart])
+	if err != nil {
+		return nil, fmt.Errorf("decode base64 header: %w", err)
+	}
+
+	var jwtHeaders map[string]interface{}
+
+	err = json.Unmarshal(headers, &jwtHeaders)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON headers: %w", err)
+	}
+
+	var payload []byte
+	if len(detachedPayload) == 0 {
+		payload, err = base64.RawURLEncoding.DecodeString(parts[jwtPayloadPart])
+		if err != nil {
+			return nil, fmt.Errorf("decode base64 payload: %w", err)
+		}
+	} else {
+		payload = detachedPayload
+	}
+
+	var jwtPayload map[string]interface{}
+
+	err = json.Unmarshal(payload, &jwtPayload)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal JSON payload: %w", err)
+	}
+
+	return &jsonWebToken{
+		headers: jwtHeaders,
+		payload: jwtPayload,
+	}, nil
+}
+
+// PrepareJWSHeaders refines input headers with JWT typ and input alg.
+func PrepareJWSHeaders(headers map[string]interface{}, alg string) map[string]interface{} {
+	newHeaders := make(map[string]interface{})
+
+	for k, v := range headers {
+		newHeaders[k] = v
+	}
+
+	newHeaders[HeaderType] = TypeJWT
+	newHeaders[HeaderAlgorithm] = alg
+
+	return newHeaders
+}
+
+// PrepareJWTUnsecuredHeaders refines input headers with JWT typ and none alg.
+func PrepareJWTUnsecuredHeaders(headers map[string]interface{}) map[string]interface{} {
+	newHeaders := make(map[string]interface{})
+
+	for k, v := range headers {
+		newHeaders[k] = v
+	}
+
+	newHeaders[HeaderType] = TypeJWT
+	newHeaders[HeaderAlgorithm] = AlgorithmNone
+
+	return newHeaders
+}
+
+// IsCompactJWS checks that input is a JWS in compact form.
+func IsCompactJWS(s string) bool {
+	parts := strings.Split(s, ".")
+
+	return len(parts) == 3 &&
```

Comment:
I assume `3` can be replaced with `jwtPartsCount`

Reply:
Right!